lib/src/Uncertainty/Algorithm/WeightedExperiments/TensorProductExperiment.cxx

```diff
@@ -0,0 +1,141 @@
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(TensorProductExperiment)
+
+static const Factory<TensorProductExperiment> Factory_TensorProductExperiment;
+
+typedef Collection< Point > PointCollection;
+
+/* Default constructor */
+TensorProductExperiment::TensorProductExperiment()
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+{
+  // Nothing to do
+}
+
+/* Constructor with parameters */
+TensorProductExperiment::TensorProductExperiment(const ExperimentPersistentCollection & collection)
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+{
+
+}
+
+/* Virtual constructor */
+TensorProductExperiment * TensorProductExperiment::clone() const
+{
+  return new TensorProductExperiment(*this);
+}
+
+/* String converter */
+String TensorProductExperiment::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << GetClassName()
+    << " name=" << getName()
+    << " collection=" << collection_;
+  return oss;
+}
+
+Bool TensorProductExperiment::hasUniformWeights() const
+{
+  return false;
+}
+
+/* Sample generation */
+Sample TensorProductExperiment::generateWithWeights(Point & weights) const
+{
+  if (!isAlreadyComputedNodesAndWeights_) computeNodesAndWeights();
+  weights = weights_;
+  return nodes_;
+}
+
+/* Compute the tensor product nodes and weights */
+void TensorProductExperiment::computeNodesAndWeights() const
+{
+  const UnsignedInteger dimension = collection_.getSize();
+  if (!(dimension > 0))
+    throw InvalidArgumentException(HERE) << "Error: expected a positive dimension, here dimension is " << dimension;
+  // Compute marginal nodes and weights
+  Collection<Sample> marginalNodesCollection;
+  Collection<Point> marginalWeightsCollection(dimension);
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    Point weights;
+    const Sample nodes(collection_[i].generateWithWeights(weights));
+    marginalNodesCollection.add(nodes);
+    marginalWeightsCollection.add(weights);
+  }
+  // Compute number of nodes for each dimension
+  // which depends on the marginal levels.
+  Indices marginal_size;
+  for (UnsignedInteger i = 0; i < dimension; ++i)
+  {
+    marginal_size[i] = marginalWeightsCollection[i].getDimension();
+  }
+  // Create index set
+  Tuples tuples(marginal_size);
+  IndicesCollection indices_collection(tuples.generate());
+  // Compute nodes and weights
+  const UnsignedInteger size = indices_collection.getSize();
+  Point weights(size);
+  Sample nodes(size, dimension);
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    weights[i] = 1.0;
+    for (UnsignedInteger j = 0; j < dimension; ++j)
+    {
+      UnsignedInteger marginal_index = indices_collection(i, j);
+      const Point marginal_weights(marginalWeightsCollection[j]);
+      weights[i] = weights[i] * marginal_weights[marginal_index];
+      Sample marginal_nodes(marginalNodesCollection[j]);
+      nodes[i][j] = marginal_nodes[marginal_index][0];
```

Comment:
nodes(i, j) and marginalNodes(marginalIndex, 0) please use the camelCase convention for the name of identifiers, variables, instance methods,g and the CamelCase convention for the class names and the static methods.

Reply:
Ok. Fixed.